src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs

```diff
@@ -91,45 +90,38 @@ internal partial class PackageInstallerService : AbstractDelayStartedService, IP
             _packageInstaller = packageInstaller;
             _packageUninstaller = packageUninstaller;
             _packageSourceProvider = packageSourceProvider;
+            ResetPackageSources();
         }
 
-        public ImmutableArray<PackageSource> GetPackageSources()
+        private void ResetPackageSources()
+            => Interlocked.Exchange(ref _packageSources, new AsyncLazy<ImmutableArray<PackageSource>>(
+                c => ComputePackageSourcesAsync(c), cacheResult: true));
```

Comment:
not 100% sure on the above.  I know that assigning to _packageSources is atomic.  However, i want to make sure that any later reads of _packageSources will see the fully created AsyncLazy.  

@jasonmalinowski do i have this right?

Reply:
You can also make _packageSources `volatile` and just assign to it here. 

---
In reply to: [414264964](https://github.com/dotnet/roslyn/pull/43630#discussion_r414264964) [](ancestors = 414264964)

Reply:
Nit: do you need the lambda? can't you just pass in `ComputePackageSourcesAsync` ?

---
In reply to: [414849964](https://github.com/dotnet/roslyn/pull/43630#discussion_r414849964) [](ancestors = 414849964,414264964)

Reply:
> @jasonmalinowski do i have this right?

Any time I have to ask myself whether the .NET framework reorders writes in it's memory model, I step back and solve the problem in a simpler way. That said, I cannot say with confidence that .NET does _not_ reorder writes and so AFAIK this is necessary.